app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java

```diff
@@ -250,80 +251,113 @@ private void onUpdatedShouldRunDecision(boolean newShouldRunDecision) {
-             */
-            if (mPollWebGuiAvailableTask == null) {
-                mPollWebGuiAvailableTask = new PollWebGuiAvailableTask(
-                        SyncthingService.this,
-                        getWebGuiUrl(),
-                        mConfig.getApiKey(),
-                        result -> {
-                    Log.i(TAG, "Web GUI has come online at " + mConfig.getWebGuiUrl());
-                    if (mApi != null) {
-                        mApi.readConfigFromRestApi();
-                    }
-                });
-            }
+        // Safety check: Log warning if a previously launched startup task did not finish properly.
+        if (mStartupTask != null && (mStartupTask.getStatus() == AsyncTask.Status.RUNNING)) {
+            Log.w(TAG, "LaunchStartupTask: StartupTask is still running. Skipped starting it twice.");
+            return;
         }
+        onServiceStateChange(State.STARTING);
+        mStartupTask = new StartupTask(this);
+        mStartupTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
     }
 
     /**
-     * Called when {@link PollWebGuiAvailableTask} confirmed the REST API is available.
-     * We can assume mApi being available under normal conditions.
+     * Sets up the initial configuration, and updates the config when coming from an old
+     * version.
+     */
+     private static class StartupTask extends AsyncTask<Void, Void, Void> {
+
+         private WeakReference<SyncthingService> refSyncthingService;
+
+         StartupTask(SyncthingService context) {
+             refSyncthingService = new WeakReference<>(context);
+         }
+
+         @Override
+         protected Void doInBackground(Void... voids) {
+             SyncthingService syncthingService = refSyncthingService.get();
+             if (syncthingService == null) {
+                 cancel(true);
+             }
+             try {
+                 syncthingService.mConfig = new ConfigXml(syncthingService);
+                 syncthingService.mConfig.updateIfNeeded();
+             } catch (ConfigXml.OpenConfigException e) {
+                 syncthingService.mNotificationHandler.showCrashedNotification(R.string.config_create_failed, true);
+                 synchronized (syncthingService.mStateLock) {
+                     syncthingService.onServiceStateChange(State.ERROR);
+                 }
+                 cancel(true);
+             }
+             return null;
+         }
+
+         @Override
+         protected void onPostExecute(Void aVoid) {
+             // Get a reference to the service if it is still there.
+             SyncthingService syncthingService = refSyncthingService.get();
+             if (syncthingService != null) {
+                 syncthingService.StartupTask_onPostExecute();
+             }
+             return;
+         }
+     }
+
+     /**
+      * Callback on {@link StartupTask#onPostExecute}.
+      */
+     private void StartupTask_onPostExecute() {
```

Comment:
Should be something like `onStartupTaskCompleteListener()`, because method names in Java are supposed to be camel case without underscores.